README.md
[![Build Status](https://travis-ci.org/artsalliancemedia/node-klv.png)](http://travis-ci.org/artsalliancemedia/node-klv)

KLV in Node
===================

A Node.js streaming parser for KLV (Key Length Value) encoded data. KLV encoding is commonly used in the motion picture industry.

http://en.wikipedia.org/wiki/KLV

KLV Streams
-----------

```javascript
var fs = require('fs'),
    klv = require('klv');

var klvStream = klv.createStream();


klvStream.on('key', function(key) {
	console.log('Key:' + key);
});

klvStream.on('value', function(value) {
	console.log('Value: ' + value);
});

klvStream.on('end', function() {
	console.log('All done');
});

var fileStream = fs.createReadStream('some_klv_file');
fileStream.pipe(klvStream);
```

Incoming streams can have multiple KLVs; the parser will emit events as it encounters each one.

Creating KLVs
-------------

Data can be encoded into a KLV; just supply the key and value as buffers and use:

```javascript
var klv = require('klv');
var myKLV = klv.encodeKLV(myKeyBuffer, myValueBuffer);
```

An optional BER length can be provided as a final parameter if you want to explicitly set the length in bytes:

```javascript
klv.encodeKLV(myKeyBuffer, myValueBuffer, 4);
```

License (MIT)
-------------

Copyright 2013 Arts Alliance Media

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

CHANGELOG.md
Release Notes
=============

0.0.7
-----

* Made passing through the data (re-emitting it) an option so the KLV stream can be used in duplex or as a through pipe

* The value parameter in encodeKLV is now optional for KLVs with no values

0.0.6
-----

* Added an option that allows the key length to be specified

* Added 'setEncoding' so that the encoding of the KLV value is automatically emitted; handy when UTF-8 isn't used

* The stream is now readable so it can be chained in pipes

0.0.5
-----

* Refactored the way KLV streams were implemented to make them more extendible; strengthened the unit tests.

0.0.4
-----

* Removed the 'length' event as it didn't serve any useful purpose.

* Removed passing through the data (re-emitting it) so the stream can be used duplex.

0.0.3
-----

* Added a BER encoder.

* Added a new function, encodeKLV, to encode a key and value buffer into a KLV.

0.0.2
-----

* Fixed compatibilty issues with Node versions < 0.8. 0.4 and 0.6 unit tests pass.

0.0.1
-----

* Initial release

test/klv_test.js
/**
 * Unit tests for KLV parsing
 */
 
var assert = require('assert'),
    stream = require('stream'),
    klv = require('../lib/klv');

describe('klv:', function() {

    describe('decodeBER', function() {
        it('should decode a BER-encoded length to an int', function() {
            // 1 byte long
            assert.equal(klv.decodeBER(new Buffer([0x00])).value, 0);
            assert.equal(klv.decodeBER(new Buffer([0x01])).value, 1);
            assert.equal(klv.decodeBER(new Buffer([0x7f])).value, 127);
            // 2 bytes long
            assert.equal(klv.decodeBER(new Buffer([0x81, 0x00])).value, 0);
            assert.equal(klv.decodeBER(new Buffer([0x81, 0x01])).value, 1);
            assert.equal(klv.decodeBER(new Buffer([0x81, 0x7f])).value, 127);
            assert.equal(klv.decodeBER(new Buffer([0x81, 0xff])).value, 255);
            // 3 bytes long
            assert.equal(klv.decodeBER(new Buffer([0x82, 0x00, 0x00])).value, 0);
            assert.equal(klv.decodeBER(new Buffer([0x82, 0x00, 0x01])).value, 1);
            assert.equal(klv.decodeBER(new Buffer([0x82, 0x00, 0x7f])).value, 127);
            assert.equal(klv.decodeBER(new Buffer([0x82, 0x01, 0x00])).value, 256);
            // 4 bytes long
            assert.equal(klv.decodeBER(new Buffer([0x83, 0x00, 0x00, 0x00])).value, 0);
            assert.equal(klv.decodeBER(new Buffer([0x83, 0x00, 0x00, 0x01])).value, 1);
            assert.equal(klv.decodeBER(new Buffer([0x83, 0x00, 0x00, 0x7f])).value, 127);
            assert.equal(klv.decodeBER(new Buffer([0x83, 0x00, 0x01, 0x00])).value, 256);
            assert.equal(klv.decodeBER(new Buffer([0x83, 0x01, 0x00, 0x00])).value, 65536);
            assert.equal(klv.decodeBER(new Buffer([0x83, 0x01, 0x00, 0x0A])).value, 65546);
            assert.equal(klv.decodeBER(new Buffer([0x83, 0x01, 0x01, 0x00])).value, 65536 + 256);
            assert.equal(klv.decodeBER(new Buffer([0x83, 0x01, 0x01, 0x01])).value, 65536 + 257);
        });
    });

    describe('encodeBER', function() {
        it('should encode an integer to a BER value', function() {
            // Single byte ints
            assert.deepEqual(klv.encodeBER(0, 4), new Buffer([0x83, 0x00, 0x00, 0x00]));
            assert.deepEqual(klv.encodeBER(10, 4), new Buffer([0x83, 0x00, 0x00, 0x0A]));
            assert.deepEqual(klv.encodeBER(127, 4), new Buffer([0x83, 0x00, 0x00, 0x7F]));
            assert.deepEqual(klv.encodeBER(255, 4), new Buffer([0x83, 0x00, 0x00, 0xFF]));
            // Multi-byte ints
            assert.deepEqual(klv.encodeBER(256, 4), new Buffer([0x83, 0x00, 0x01, 0x00]));
            assert.deepEqual(klv.encodeBER(513, 4), new Buffer([0x83, 0x00, 0x02, 0x01]));
            assert.deepEqual(klv.encodeBER(65535, 4), new Buffer([0x83, 0x00, 0xFF, 0xFF]));
            assert.deepEqual(klv.encodeBER(65536, 4), new Buffer([0x83, 0x01, 0x00, 0x00]));
            assert.deepEqual(klv.encodeBER(65546, 4), new Buffer([0x83, 0x01, 0x00, 0x0A]));
            assert.deepEqual(klv.encodeBER(65536 + 256, 4), new Buffer([0x83, 0x01, 0x01, 0x00]));
            assert.deepEqual(klv.encodeBER(65536 + 257, 4), new Buffer([0x83, 0x01, 0x01, 0x01]));
        });
    });

    describe('encode/decodeBER', function() {
        it('should encode and decode a BER, producing the same value', function() {
            [0, 10, 127, 256, 513, 65535, 65536, 65546, 104532, 256*256*256*24].forEach(function(value) {
                assert.equal(value, klv.decodeBER(klv.encodeBER(value)).value);
            });
        });
    });

    describe('createStream#single KLV', function() {
        it('should create a new KLV stream and transform the incoming stream into a KLV', function(done) {
            var testKLV = new Buffer([
                  0x03, 0x2E, 0x5F, 0xAB, 0x08, 0x12, 0x2F, 0x0C, 0xEE, 0x33, 0x00, 0x01, 0x02, 0x45, 0x6D, 0xDD,
                  0x83, 0x00, 0x00, 0x05,
                  0x05, 0x04, 0x03, 0x02, 0x01])

            var valueEventReceived = false;

            // Create a simple read stream for the testKLV bytes
            var byteStream = new stream.Stream();
            byteStream.readable = true;

            var klvStream = klv.createStream();
            byteStream.pipe(klvStream);

            klvStream.on('key', function(key) {
                assert(key);
                assert.equal(key.length, 16);
                assert.deepEqual(key, testKLV.slice(0, 16));
            });

            klvStream.on('value', function(value) {
                assert(value);
                assert.equal(typeof value, 'object');
                assert.equal(value.length, 5);
                assert.deepEqual(value, testKLV.slice(20));
                valueEventReceived = true;
            });

            klvStream.on('end', function() {
                assert(valueEventReceived);
                done();
            });

            // Emit the KLV 
            byteStream.emit('data', testKLV);
            byteStream.emit('end');
        });
    });

    describe('createStream#single KLV with custom key length', function() {
        it('should create a new KLV stream with custom key kength and transform the incoming stream into a KLV', function(done) {
            
            var testKLV = new Buffer([
                // Key length of 18
                0x03, 0x2E, 0x5F, 0xAB, 0x08, 0x12, 0x2F, 0x0C, 0xEE, 0x33, 0x00, 0x01, 0x02, 0x45, 0x6D, 0xDD, 0x55, 0x67,
                0x83, 0x00, 0x00, 0x05,
                0x05, 0x04, 0x03, 0x02, 0x01])

            var valueEventReceived = false;

            // Create a simple read stream for the testKLV bytes
            var byteStream = new stream.Stream();
            byteStream.readable = true;

            var klvStream = klv.createStream({keyLength: 18});
            byteStream.pipe(klvStream);

            klvStream.on('key', function(key) {
                assert(key);
                assert.equal(key.length, 18);
                assert.deepEqual(key, testKLV.slice(0, 18));
            });

            klvStream.on('value', function(value) {
                assert(value);
                assert.equal(typeof value, 'object');
                assert.equal(value.length, 5);
                assert.deepEqual(value, testKLV.slice(22));
                valueEventReceived = true;
            });

            klvStream.on('end', function() {
                assert(valueEventReceived);
                done();
            });

            // Emit the KLV 
            byteStream.emit('data', testKLV);
            byteStream.emit('end');
        });
    });

    describe('createStream#multiple KLVs', function() {
        it('should create a new KLV stream and transform the incoming stream into a KLV', function(done) {
            
            var testKLV1 = new Buffer([
                  0x03, 0x2E, 0x5F, 0xAB, 0x08, 0x12, 0x2F, 0x0C, 0xEE, 0x33, 0x00, 0x01, 0x02, 0x45, 0x6D, 0xDD,
                  0x83, 0x00, 0x00, 0x05,
                  0x05, 0x04, 0x03, 0x02, 0x01]);
            var testKLV2 = new Buffer([
                  0xFF, 0x3D, 0x99, 0x23, 0x4D, 0x01, 0x02, 0x03, 0x04, 0xAD, 0xAC, 0xAB, 0xAA, 0x00, 0x00, 0x00,
                  0x82, 0x00, 0x07,
                  0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07]);

            // Create a simple read stream for the testKLV bytes
            var byteStream = new stream.Stream();
            byteStream.readable = true;

            var klvStream = klv.createStream();
            byteStream.pipe(klvStream);

            var iterationCounter = 1;

            klvStream.on('key', function(key) {
                assert(key);
                assert.equal(key.length, 16);
                if (iterationCounter === 1)
                    assert.deepEqual(key, testKLV1.slice(0, 16));
                else
                    assert.deepEqual(key, testKLV2.slice(0, 16));
            });
    
            klvStream.on('value', function(value) {
                assert(value);
                assert.equal(typeof value, 'object');
                if (iterationCounter === 1) {
                    assert.equal(value.length, 5);
                    assert.deepEqual(value, testKLV1.slice(20, 25));
                }
                else {
                    assert.equal(value.length, 7);
                    assert.deepEqual(value, testKLV2.slice(19, 26));
                }
                iterationCounter++;
            });

            klvStream.on('end', function() {
                assert.equal(iterationCounter, 3);
                done();
            });

            // Emit the KLVs in multiple chopped up data events
            byteStream.emit('data', testKLV1.slice(0,10));
            byteStream.emit('data', Buffer.concat([testKLV1.slice(10), testKLV2.slice(0, 4)]));
            byteStream.emit('data', testKLV2.slice(4));
            byteStream.emit('end');
        });
    });

    describe('encodeKLV', function() {
        it('should encode a key and a value into a KLV', function() {
            var testKey = new Buffer([0x03, 0x2E, 0x5F, 0xAB, 0x08, 0x12, 0x2F, 0x0C, 0xEE, 0x33, 0x00, 0x01, 0x02, 0x45, 0x6D, 0xDD]);
            var testValue = new Buffer([0x05, 0x04, 0x03, 0x02, 0x01]);
            var encodedKLV = klv.encodeKLV(testKey, testValue);
            assert.deepEqual(encodedKLV, Buffer.concat([testKey, new Buffer([0x05]), testValue]));
            var encodedKLV = klv.encodeKLV(testKey, testValue, 4);
            assert.deepEqual(encodedKLV, Buffer.concat([testKey, new Buffer([0x83, 0x00, 0x00, 0x05]), testValue]));
        });
    });

    describe('encodeKLV with different key length to 16', function() {
        it('should encode a key (of length 18) and a value into a KLV', function() {
            var testKey = new Buffer([0x03, 0x2E, 0x5F, 0xAB, 0x08, 0x12, 0x2F, 0x0C, 0xEE,
                                      0x33, 0x00, 0x01, 0x02, 0x45, 0x6D, 0xDD, 0x55, 0x96]);
            var testValue = new Buffer([0x05, 0x04, 0x03, 0x02, 0x01]);
            var encodedKLV = klv.encodeKLV(testKey, testValue);
            assert.deepEqual(encodedKLV, Buffer.concat([testKey, new Buffer([0x05]), testValue]));
            var encodedKLV = klv.encodeKLV(testKey, testValue, 4);
            assert.deepEqual(encodedKLV, Buffer.concat([testKey, new Buffer([0x83, 0x00, 0x00, 0x05]), testValue]));
        });
    });

    describe('setEncoding', function() {
        it('should output KLV values in the proper encoded string format', function() {
            var testKey = new Buffer([0x03, 0x2E, 0x5F, 0xAB, 0x08, 0x12, 0x2F, 0x0C, 0xEE, 0x33, 0x00, 0x01, 0x02, 0x45, 0x6D, 0xDD]);
            var testValue = new Buffer('these are little endian encoded Unicode characters', 'ucs2');
            var encodedKLV = klv.encodeKLV(testKey, testValue);
            assert.deepEqual(encodedKLV, Buffer.concat([testKey, new Buffer([0x64]), testValue]));
            var encodedKLV = klv.encodeKLV(testKey, testValue, 4);
            assert.deepEqual(encodedKLV, Buffer.concat([testKey, new Buffer([0x83, 0x00, 0x00, 0x64]), testValue]));

            // test outputting
            var klvStream = klv.createStream();
            klvStream.setEncoding('ucs2');
            klvStream.on('value', function(value) {
                assert.equal(value.length, 50);
                assert.equal(value, 'these are little endian encoded Unicode characters');
            });
            klvStream.write(encodedKLV);
            klvStream.end();
        });
    });

});

lib/buffer.js
/**
 * Shim for adding concat function to Buffer for versions of Node < 0.8
 *
 * https://github.com/polotek/procstreams/blob/master/compat/buffer.js
 */

if(!Buffer.concat) {

  Buffer.concat = function(list, length) {
    if (!Array.isArray(list)) {
      throw new Error('Usage: Buffer.concat(list, [length])');
    }

    if (list.length === 0) {
      return new Buffer(0);
    } else if (list.length === 1) {
      return list[0];
    }

    if (typeof length !== 'number') {
      length = 0;
      for (var i = 0; i < list.length; i++) {
        var buf = list[i];
        length += buf.length;
      }
    }

    var buffer = new Buffer(length);
    var pos = 0;
    for (var i = 0; i < list.length; i++) {
      var buf = list[i];
      buf.copy(buffer, pos);
      pos += buf.length;
    }
    return buffer;
  };

}

lib/klv.js
/**
 * KLV (Key Value Length) parser
 * http://en.wikipedia.org/wiki/KLV
 */
var stream = require('stream'),
    util = require('util');

var KLV_STATES = {KEY: 0, LENGTH: 1, VALUE: 2};

/**
 * Create and return a new KLVStream
 */
exports.createStream = function(options) {
    return new KLVStream(options);
};

/**
 * Creates a KLV streaming parser. Options are:
 *
 * keyLength: specify the length of the KLV key
 * emitData: if true then values will be emitted as data events
 */
var KLVStream = function(options) {
    this.writable = true;
    this.readable = true;

    // Default key length is 16
    this.keyLength = (options && options.keyLength) || 16;

    // Check whether KLV values should be emitted as data
    this.emitData = Boolean(options && options.emitData);

    this.buffer = new Buffer(0);
    this.state = KLV_STATES.KEY;
    this.valueLength = 0;
};

util.inherits(KLVStream, stream.Stream);

/**
 * Sets the encoding type for the KLV value
 */
KLVStream.prototype.setEncoding = function(encoding) {
    this.encoding = encoding;
    return this;
};

KLVStream.prototype.write = function(data) {
    this.buffer = Buffer.concat([this.buffer, data]);
    // Iterate through the buffer, pulling out the KLV components
    // There may be multiple KLVs in the buffer
    var done = false;
    while(!done) {
        // Emit the key
        if (this.state === KLV_STATES.KEY && this.buffer.length >= this.keyLength) {
            var key = this.buffer.slice(0, this.keyLength);
            this.buffer = this.buffer.slice(this.keyLength); // Remove the key portion from the buffer
            this.state = KLV_STATES.LENGTH;
            this.emit('key', key);
        }
        // Emit the length
        else if (this.state === KLV_STATES.LENGTH) {
            var decodedBER = decodeBER(this.buffer);
            this.buffer = this.buffer.slice(decodedBER.bytesRead);
            this.state = KLV_STATES.VALUE;
            this.valueLength = decodedBER.value; // Record to track the value
            this.emit('length', this.valueLength);
        }

        // Emit the value, using the correct encoding if specified
        else if (this.state === KLV_STATES.VALUE && this.buffer.length >= this.valueLength) {
            var value = this.buffer.slice(0, this.valueLength);
            this.buffer = this.buffer.slice(this.valueLength);
            this.state = KLV_STATES.KEY;
            // Use the encoding if specified
            this.emit('value', (this.encoding) ? value.toString(this.encoding) : value);
            if (this.emitData)
                this.emit('data', (this.encoding) ? value.toString(this.encoding) : value);
        }
        else
            done = true;
    }
    return true;
};

KLVStream.prototype.end = function() {
    this.emit('end');
};

/**
 * Encodes a key and value buffer into a KLV
 * The length in bytes of the BER encoded length can be optionally specified
 */
exports.encodeKLV = function(key, value, berLength) {
    // Value can be null or undefined
    if (!value)
        value = new Buffer(0);
    // Convert arrays to buffers if necessary
    if (util.isArray(key))
        key = new Buffer(key);
    if (util.isArray(value))
        value = new Buffer(value);
    return Buffer.concat([key, encodeBER(value.length, berLength), value]);
};

/**
 * Decodes a BER integer value
 * See http://en.wikipedia.org/wiki/KLV for more info, but to paraphrase:
 *
 * If the first byte in the BER length field does not have the high bit set (0x80),
 * then that single byte represents an integer between 0 and 127 and indicates
 * the number of Value bytes that immediately follows. If the high bit is set,
 * then the lower seven bits indicate how many bytes follow that themselves make up a length field.
 *
 * Bytes should be passed in as a Buffer
 */
var decodeBER = exports.decodeBER = function(buffer) {
    var length = buffer[0];
    var bytesRead = 1;
    if (length > 127) {
        bytesRead += length & 127; // Strip off the high bit
        length = 0;
        for (var i = 1; i < bytesRead; i++) {
            length += buffer[i] << (8 * (bytesRead - i - 1));
        }
    }
    return {value: length, bytesRead: bytesRead};
};

/**
 * Encodes an integer to BER
 *
 * The length of the encoded BER value (in bytes) can be optionally specified.
 */
var encodeBER = exports.encodeBER = function(value, berLength) {
    if (!berLength) {
        if (value < 127)
            return new Buffer([value]);
        else if (value < 256)
            berLength = 2;
        else if (value < 256 * 256)
            berLength = 3;
        else if (value < 256 * 256 * 256)
            berLength = 4;
        else
            berLength = 5; // 32 bit unsigned int is the max for this function
    }
    // Add the BER byte length
    var ber = [127 + berLength];
    for (var i = 1; i < berLength; i++)
        ber[i] = (value >> (8 * (berLength - i - 1))) & 255
    return new Buffer(ber);
};